service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -12,90 +12,110 @@
 import org.springframework.transaction.annotation.Transactional;
 
-import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
-import java.util.stream.Collectors;
 
 import static com.epam.esm.model.parameters.CustomErrorCode.*;
 
 @Service
+@AllArgsConstructor
 public class CertificateServiceImpl implements CertificateService {
+    @Autowired
     private final GiftCertificateDao certificateDao;
-    private final TagDao tagDao;
-
     @Autowired
-    public CertificateServiceImpl(GiftCertificateDao certificateDao, TagDao tagDao) {
-        this.certificateDao = certificateDao;
-        this.tagDao = tagDao;
-    }
+    private final TagDao tagDao;
 
     @Override
     @Transactional
     public GiftCertificateDto add(GiftCertificateDto giftCertificate) throws RecourseExistException {
-        Map<String, String> groupParams = new HashMap<>();
         Optional<GiftCertificateDto> tempGiftCertificate = certificateDao.findByName(giftCertificate.getName());
         if (tempGiftCertificate.isPresent()) {
-            throw new RecourseExistException(CustomErrorCode.RECOURSE_CERTIFICATE_EXIST,"Certificate={" + giftCertificate.toString() + "} exists");
+            throw new RecourseExistException(CustomErrorCode.RECOURSE_CERTIFICATE_EXIST, "Certificate={" + giftCertificate.toString() + "} exists");
         }
-        long certificateId = certificateDao.create(giftCertificate).getId();
-        for (TagDto tagDto : giftCertificate.getListTag()
+        for (TagDto tagDto : giftCertificate.getListTagDto()
         ) {
             Optional<TagDto> tempTag = tagDao.findByName(tagDto.getTagName());
             tagDto.setTagId(tempTag.map(TagDto::getTagId).orElseGet(() -> tagDao.create(tagDto).getTagId()));
         }
-        List<TagDto> distinctListTagDto = giftCertificate.getListTag().stream().distinct().collect(Collectors.toList());
-        for (TagDto tagDto : distinctListTagDto
-        ) {
-            certificateDao.addLinkTag(certificateId, tagDto.getTagId());
-        }
-        return certificateDao.findEntityById(certificateId).get();
+        return certificateDao.create(giftCertificate);
     }
 
     @Override
     @Transactional
     public GiftCertificateDto update(GiftCertificateDto giftCertificate) throws RecourseNotExistException {
         Optional<GiftCertificateDto> optionalGiftCertificate = certificateDao.findEntityById(giftCertificate.getId());
         if (!optionalGiftCertificate.isPresent()) {
-            throw new RecourseNotExistException(RECOURSE_CERTIFICATE_NOT_EXIST,"Certificate with id={" + giftCertificate.getId() + "} doesn't exist");
+            throw new RecourseNotExistException(RECOURSE_CERTIFICATE_NOT_EXIST, "Certificate with id={" + giftCertificate.getId() + "} doesn't exist");
         }
-        return certificateDao.update(giftCertificate);
+        return certificateDao.update(merge(giftCertificate, optionalGiftCertificate.get()));
```

Comment:
Just some notes for the future. No needs to fix.
There is `IFNULL` SQL  method that can be used instead of `merge()` method here to update only the fields that should be updated.

Reply:
Noted